### PR TITLE
Refactor DockerClient and ImageManager from DevcontainerRunner

### PR DESCRIPTION
**Pull request by artmoskvin**
TODO:

- [x] Refactor Random from ImageManager
- [x] Refactor RegistryCredentials from ImageManager
- [x] Refactor OutputPrinter from ImageManager
- [x] Add more tests for ImageManager
- [ ] Add more tests for Runner

### CODE REVIEW

File: pkg/devcontainer/docker_client.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+import (
+	"context"
+	"io"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/image"
+	"github.com/docker/docker/api/types/network"
+	"github.com/docker/docker/client"
```
Comment (aleh-null) [inline pkg/devcontainer/docker_client.go:11]: hm, seems like it's now `github.com/moby/moby`

File: pkg/devcontainer/docker_client.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+	v1 "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+type DockerClient interface {
```
Comment (aleh-null) [inline pkg/devcontainer/docker_client.go:15]: I think we should just use [this](https://github.com/moby/moby/blob/master/client/interface.go#L48)

File: pkg/devcontainer/image_manager.go [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+	"github.com/artmoskvin/hide/pkg/util"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/image"
+	"github.com/docker/docker/api/types/registry"
+	"github.com/docker/docker/pkg/archive"
+	"github.com/rs/zerolog/log"
+)
+
+type ImageManager interface {
```
Comment (aleh-null) [inline pkg/devcontainer/image_manager.go:15]: Let's use [this](https://github.com/moby/moby/blob/master/client/interface.go#L89)

File: pkg/devcontainer/docker_client.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+	ImagePull(ctx context.Context, refStr string, options image.PullOptions) (io.ReadCloser, error)
+	ImageBuild(ctx context.Context, buildContext io.Reader, options types.ImageBuildOptions) (types.ImageBuildResponse, error)
+}
+
+type DefaultDockerClient struct {
+	*client.Client
+}
+
+func NewDefaultDockerClient() (DockerClient, error) {
+	cli, err := client.NewClientWithOpts(client.FromEnv)
```
Comment (aleh-null) [inline pkg/devcontainer/docker_client.go:31]: How should we think about ENV in context of other options and config for Hide?

File: pkg/devcontainer/image_manager.go [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+type DockerImageManager struct {
+	dockerClient DockerClient
+	config       DockerRunnerConfig
+}
+
+func NewImageManager(dockerClient DockerClient, config DockerRunnerConfig) ImageManager {
+	return &DockerImageManager{dockerClient: dockerClient, config: config}
+}
+
+func (im *DockerImageManager) PullOrBuildImage(ctx context.Context, workingDir string, config Config) (string, error) {
```
Comment (aleh-null) [inline pkg/devcontainer/image_manager.go:31]: I thought you did not like `Or` methods. What is the rational for it?

File: pkg/devcontainer/mocks/mock_docker_client.go [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/image"
+	"github.com/docker/docker/api/types/network"
+	v1 "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+// MockDockerClient is a mock implementation of the DockerClient interface for testing
+type MockDockerClient struct {
```
Comment (aleh-null) [inline pkg/devcontainer/mocks/mock_docker_client.go:15]: Lets try [this](https://github.com/samalba/dockerclient/blob/master/mockclient/mock.go#L10)

File: pkg/devcontainer/runner.go [hunk 7]
```diff
@@ -41,28 +32,28 @@ type ExecResult struct {
 }
 
 type Runner interface {
-	Run(projectPath string, config Config) (string, error)
-	Stop(containerId string) error
-	Exec(containerId string, command []string) (ExecResult, error)
+	Run(ctx context.Context, projectPath string, config Config) (string, error)
```
Comment (aleh-null) [inline pkg/devcontainer/runner.go:30]: Really nice with context.

**Review by aleh-null — COMMENTED** [review]
Not sure I completely follow DockerRunner, let's have a chat.

File: pkg/devcontainer/image_manager.go (hunk 5 above)
Comment (artmoskvin) [reply pkg/devcontainer/image_manager.go:31]: TODO:
- split into separate methods
- create aux methods in devcon.Config to define config type

File: pkg/devcontainer/mocks/mock_docker_client.go (hunk 6 above)
Comment (artmoskvin) [reply pkg/devcontainer/mocks/mock_docker_client.go:19]: Do or do not. There is no try ©

File: pkg/devcontainer/image_manager.go (hunk 5 above)
Comment (artmoskvin) [reply pkg/devcontainer/image_manager.go]: Done

File: pkg/devcontainer/mocks/mock_docker_client.go (hunk 6 above)
Comment (artmoskvin) [reply pkg/devcontainer/mocks/mock_docker_client.go:19]: Done

File: pkg/devcontainer/image_manager.go (hunk 3 above)
Comment (artmoskvin) [reply pkg/devcontainer/image_manager.go:16]: Done

File: pkg/devcontainer/docker_client.go (hunk 2 above)
Comment (artmoskvin) [reply pkg/devcontainer/docker_client.go:15]: Done

File: pkg/devcontainer/docker_client.go [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+	ContainerStart(ctx context.Context, containerID string, options container.StartOptions) error
+	ContainerStop(ctx context.Context, containerID string, options container.StopOptions) error
+	ContainerExecCreate(ctx context.Context, container string, config types.ExecConfig) (types.IDResponse, error)
+	ContainerExecAttach(ctx context.Context, execID string, config types.ExecStartCheck) (types.HijackedResponse, error)
+	ContainerExecInspect(ctx context.Context, execID string) (types.ContainerExecInspect, error)
+	ImagePull(ctx context.Context, refStr string, options image.PullOptions) (io.ReadCloser, error)
+	ImageBuild(ctx context.Context, buildContext io.Reader, options types.ImageBuildOptions) (types.ImageBuildResponse, error)
+}
+
+type DefaultDockerClient struct {
```
[the remark this answers is not in the document]
Comment (artmoskvin) [reply pkg/devcontainer/docker_client.go]: We don't need this struct anymore. I removed it.

File: pkg/devcontainer/docker_client.go [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+	*client.Client
+}
+
+func NewDefaultDockerClient() (DockerClient, error) {
```
[the remark this answers is not in the document]
Comment (artmoskvin) [reply pkg/devcontainer/docker_client.go]: We don't need this struct anymore. I removed it.

File: pkg/devcontainer/docker_client.go (hunk 1 above)
Comment (artmoskvin) [reply pkg/devcontainer/docker_client.go]: It's complicated :)

**Comment by artmoskvin** [comment]
When running tests locally I get this error:
```
=== RUN   TestDockerImageManager_BuildImage/Build_image
{"level":"debug","buildContextPath":"/workdir","time":"2024-08-29T15:41:30+02:00","message":"Building image"}
==================
WARNING: DATA RACE
Write at 0x00c00012e300 by goroutine 9:
  ??()
      -:0 +0x1026c7688
  sync/atomic.CompareAndSwapInt32()
      <autogenerated>:1 +0x18
  io.(*pipe).write()
      /opt/homebrew/Cellar/go/1.23.0/libexec/src/io/pipe.go:81 +0x70
  io.(*PipeWriter).Write()
      /opt/homebrew/Cellar/go/1.23.0/libexec/src/io/pipe.go:161 +0x44
  bufio.(*Writer).Flush()
      /opt/homebrew/Cellar/go/1.23.0/libexec/src/bufio/bufio.go:639 +0xc0
  github.com/docker/docker/pkg/archive.CompressStream.(*BufioWriterPool).NewWriteCloserWrapper.func1()
      /Users/artemm/go/pkg/mod/github.com/docker/docker@v26.1.3+incompatible/pkg/pools/pools.go:130 +0x44
  github.com/docker/docker/pkg/ioutils.(*writeCloserWrapper).Close()
      /Users/artemm/go/pkg/mod/github.com/docker/docker@v26.1.3+incompatible/pkg/ioutils/writers.go:43 +0x58
  github.com/docker/docker/pkg/archive.(*Tarballer).Do.func1()
      /Users/artemm/go/pkg/mod/github.com/docker/docker@v26.1.3+incompatible/pkg/archive/archive.go:919 +0x108
  runtime.deferreturn()
      /opt/homebrew/Cellar/go/1.23.0/libexec/src/runtime/panic.go:605 +0x5c
  github.com/docker/docker/pkg/archive.TarWithOptions.gowrap1()
      /Users/artemm/go/pkg/mod/github.com/docker/docker@v26.1.3+incompatible/pkg/archive/archive.go:849 +0x34

Previous read at 0x00c00012e300 by goroutine 8:
  reflect.typedmemmove()
      /opt/homebrew/Cellar/go/1.23.0/libexec/src/runtime/mbarrier.go:225 +0x0
  reflect.packEface()
      /opt/homebrew/Cellar/go/1.23.0/libexec/src/reflect/value.go:133 +0xa4
  reflect.valueInterface()
      /opt/homebrew/Cellar/go/1.23.0/libexec/src/reflect/value.go:1510 +0x148
  reflect.Value.Interface()
      /opt/homebrew/Cellar/go/1.23.0/libexec/src/reflect/value.go:1481 +0x74
  fmt.(*pp).printValue()
      /opt/homebrew/Cellar/go/1.23.0/libexec/src/fmt/print.go:769 +0x80
  fmt.(*pp).printValue()
      /opt/homebrew/Cellar/go/1.23.0/libexec/src/fmt/print.go:921 +0xf70
  fmt.(*pp).printArg()
      /opt/homebrew/Cellar/go/1.23.0/libexec/src/fmt/print.go:759 +0x8a0
  fmt.(*pp).doPrintf()
      /opt/homebrew/Cellar/go/1.23.0/libexec/src/fmt/print.go:1173 +0xc58
  fmt.Sprintf()
      /opt/homebrew/Cellar/go/1.23.0/libexec/src/fmt/print.go:239 +0x50
  github.com/stretchr/testify/mock.Arguments.Diff()
      /Users/artemm/go/pkg/mod/github.com/stretchr/testify@v1.9.0/mock/mock.go:939 +0x13c
  github.com/stretchr/testify/mock.(*Mock).findExpectedCall()
      /Users/artemm/go/pkg/mod/github.com/stretchr/testify@v1.9.0/mock/mock.go:368 +0x118
  github.com/stretchr/testify/mock.(*Mock).MethodCalled()
      /Users/artemm/go/pkg/mod/github.com/stretchr/testify@v1.9.0/mock/mock.go:476 +0x6c
  github.com/stretchr/testify/mock.(*Mock).Called()
      /Users/artemm/go/pkg/mod/github.com/stretchr/testify@v1.9.0/mock/mock.go:466 +0x150
  github.com/artmoskvin/hide/pkg/devcontainer/mocks.(*MockDockerImageClient).ImageBuild()
      /Users/artemm/Code/hide/pkg/devcontainer/mocks/mock_docker_image_client.go:22 +0x180
  github.com/artmoskvin/hide/pkg/devcontainer.(*DockerImageManager).BuildImage()
      /Users/artemm/Code/hide/pkg/devcontainer/image_manager.go:117 +0x8b8
  github.com/artmoskvin/hide/pkg/devcontainer_test.TestDockerImageManager_BuildImage.func10()
      /Users/artemm/Code/hide/pkg/devcontainer/image_manager_test.go:247 +0x2b8
  testing.tRunner()
      /opt/homebrew/Cellar/go/1.23.0/libexec/src/testing/testing.go:1690 +0x184
  testing.(*T).Run.gowrap1()
      /opt/homebrew/Cellar/go/1.23.0/libexec/src/testing/testing.go:1743 +0x40

Goroutine 9 (running) created at:
  github.com/docker/docker/pkg/archive.TarWithOptions()
      /Users/artemm/go/pkg/mod/github.com/docker/docker@v26.1.3+incompatible/pkg/archive/archive.go:849 +0xa4
  github.com/artmoskvin/hide/pkg/devcontainer.(*DockerImageManager).BuildImage()
      /Users/artemm/Code/hide/pkg/devcontainer/image_manager.go:90 +0x3b4
  github.com/artmoskvin/hide/pkg/devcontainer_test.TestDockerImageManager_BuildImage.func10()
      /Users/artemm/Code/hide/pkg/devcontainer/image_manager_test.go:247 +0x2b8
  testing.tRunner()
      /opt/homebrew/Cellar/go/1.23.0/libexec/src/testing/testing.go:1690 +0x184
  testing.(*T).Run.gowrap1()
      /opt/homebrew/Cellar/go/1.23.0/libexec/src/testing/testing.go:1743 +0x40

Goroutine 8 (running) created at:
  testing.(*T).Run()
      /opt/homebrew/Cellar/go/1.23.0/libexec/src/testing/testing.go:1743 +0x5e0
  github.com/artmoskvin/hide/pkg/devcontainer_test.TestDockerImageManager_BuildImage()
      /Users/artemm/Code/hide/pkg/devcontainer/image_manager_test.go:241 +0x5ec
  testing.tRunner()
      /opt/homebrew/Cellar/go/1.23.0/libexec/src/testing/testing.go:1690 +0x184
  testing.(*T).Run.gowrap1()
      /opt/homebrew/Cellar/go/1.23.0/libexec/src/testing/testing.go:1743 +0x40
==================
{"level":"debug","tag":"test:latest","time":"2024-08-29T15:41:30+02:00","message":"Built image"}
    testing.go:1399: race detected during execution of test
    --- FAIL: TestDockerImageManager_BuildImage/Build_image (0.00s)
```

**Comment by artmoskvin** [comment]
I figured out why I could reproduce this bug locally. Turns out that my neovim plugin for tests uses the `-race` flag. If you run `go test -race ./...` from your shell you'll get the same error. I wonder if it's a good idea to add this flag in our CI/CD pipeline @aleh-null ?

**Review by aleh-null — APPROVED** [review]

**Pull request merged**